File: src/stores/reports.ts
import { create } from 'zustand';
import { toast } from 'sonner';
import { getTransport } from '../lib/transport';

// =====================================================================
// Types — mirror crates/atomic-core/src/models.rs
// =====================================================================

/// Discriminator on every atom; reports return findings with kind=report.
export type AtomKind = 'captured' | 'report';

/// Wire shape of `SourceScopeWindow`. The Rust enum is
/// `#[serde(rename_all = "snake_case")]` over `{ SinceLastRun,
/// Duration(String) }`, which externally-tagged serializes as either
/// the bare string `"since_last_run"` or the object `{"duration":
/// "P7D"}`. The TS shape must match exactly or the backend rejects the
/// payload during JSON deserialization.
export type SourceScopeWindow = 'since_last_run' | { duration: string };

/// Wire shape of `ContextScopeWindow`. Distinct enum from the source
/// window — the contradiction-scan idiom needs `OlderThanSource` which
/// has no source-side analog.
export type ContextScopeWindow = 'older_than_source' | { duration: string };

/// Wire-aligned `ContextScopeMode`. Backend variants are
/// `same_as_source | all | explicit`. "Explicit" with an empty tag
/// list is how the UI expresses "no context" — there is no `None`
/// variant.
export type ContextScopeMode = 'same_as_source' | 'all' | 'explicit';

/// Wire-aligned `CitationPolicy`. Backend variants are `source_only`
/// (citations resolve to atoms in the run's source scope) and
/// `source_and_context` (semantic_search results also become citable).
export type CitationPolicy = 'source_only' | 'source_and_context';

/// One report definition. Matches `Report` in atomic-core. Cache fields
/// (`last_run_at`, `last_finding_atom_id`, `last_error`) are advisory —
/// authoritative state lives on `task_runs` + `report_findings`.
export interface Report {
  id: string;
  name: string;
  description: string | null;
  research_prompt: string;

  source_scope_tag_ids: string[];
  source_scope_window: SourceScopeWindow | null;
  source_include_kinds: AtomKind[];

  context_scope_mode: ContextScopeMode;
  context_scope_tag_ids: string[];
  context_scope_window: ContextScopeWindow | null;
  context_include_kinds: AtomKind[];

  citation_policy: CitationPolicy;

  max_source_atoms: number | null;
  max_source_tokens: number | null;
  max_tool_iterations: number | null;

  schedule: string;
  schedule_tz: string | null;

  enabled: boolean;
  output_atom_tags: string[];

  last_run_at: string | null;
  last_finding_atom_id: string | null;
  last_error: string | null;

  created_at: string;
  updated_at: string;
}

export interface ReportFinding {
  finding_atom_id: string;
  report_id: string | null;
  run_id: string | null;
  report_name_snapshot: string;
  created_at: string;
}

export interface ReportFindingCitation {
  finding_atom_id: string;
  cited_atom_id: string;
  position: number;
  excerpt: string;
}

/// Cleaned-up shape we cache after destructuring the
/// `list_findings_for_report` response. The wire format is a JSON
/// 2-tuple `[ReportFinding, AtomWithTags]` (because Rust serializes
/// `Vec<(A, B)>` as `[[a,b], ...]`); the store decodes those tuples
/// into this object before handing them to consumers. AtomWithTags
/// uses `#[serde(flatten)]` so atom fields sit at the top of `atom`.
export interface ReportFindingWithAtom {
  finding: ReportFinding;
  atom: {
    id: string;
    content: string;
    source_url: string | null;
    created_at: string;
    updated_at: string;
    kind: AtomKind;
    [k: string]: unknown;
  };
}

/// Raw wire shape — exactly what the server returns for
/// `list_findings_for_report`. Kept private so consumers see the clean
/// object form above.
type FindingTuple = [ReportFinding, ReportFindingWithAtom['atom']];

// =====================================================================
// Write request shapes — mirror Create/UpdateReportRequest
// =====================================================================

/// `POST /api/reports` body. Mirrors `CreateReportRequest`. Backend
/// fills sensible defaults for any field omitted; we keep them
/// optional here for the editor's progressive-disclosure UX.
export interface CreateReportInput {
  name: string;
  description?: string | null;
  research_prompt: string;

  source_scope_tag_ids?: string[];
  source_scope_window?: SourceScopeWindow | null;
  source_include_kinds?: AtomKind[];

  context_scope_mode?: ContextScopeMode;
  context_scope_tag_ids?: string[];
  context_scope_window?: ContextScopeWindow | null;
  context_include_kinds?: AtomKind[];

  citation_policy?: CitationPolicy;

  max_source_atoms?: number | null;
  max_source_tokens?: number | null;
  max_tool_iterations?: number | null;

  schedule: string;
  schedule_tz?: string | null;

  enabled?: boolean;
  output_atom_tags?: string[];
}

/// `PUT /api/reports/:id` body. Every field optional; only present
/// fields are written. Mirrors `UpdateReportRequest`. Note the nested
/// `Option<Option<T>>` pattern from Rust collapses to plain optional
/// here — we send `null` when the user is explicitly clearing the
/// field, omit the key entirely when they aren't touching it.
export interface UpdateReportInput {
  name?: string;
  description?: string | null;
  research_prompt?: string;

  source_scope_tag_ids?: string[];
  source_scope_window?: SourceScopeWindow | null;
  source_include_kinds?: AtomKind[];

  context_scope_mode?: ContextScopeMode;
  context_scope_tag_ids?: string[];
  context_scope_window?: ContextScopeWindow | null;
  context_include_kinds?: AtomKind[];

  citation_policy?: CitationPolicy;

  max_source_atoms?: number | null;
  max_source_tokens?: number | null;
  max_tool_iterations?: number | null;

  schedule?: string;
  schedule_tz?: string | null;

  enabled?: boolean;
  output_atom_tags?: string[];
}

// =====================================================================
// Store
// =====================================================================

interface ReportsStore {
  reports: Report[];
  byId: Record<string, Report>;

  /// Cached last finding per report so the list view's tertiary line
  /// (the italic excerpt) doesn't issue N requests on every render.
  /// `null` after a fetch attempt means "no findings yet" — distinguishes
  /// from `undefined` ("never fetched").
  lastFindingByReport: Record<string, ReportFindingWithAtom | null>;

  /// Full findings cache, keyed by reportId. Populated by the detail
  /// view on mount. Most-recent first; the order matches the wire
  /// response from `list_findings_for_report`.
  findingsByReport: Record<string, ReportFindingWithAtom[]>;

  /// Lazy-loaded citation counts keyed by finding atom id. The
  /// findings response carries the finding row + atom but not the
  /// citation count, so the detail-view row mounts a small effect
  /// that calls `list_finding_citations` and stashes the count here.
  /// `undefined` = not yet fetched, number = resolved.
  citationCountsByAtomId: Record<string, number>;

  /// Reports the user has manually dispatched and whose run hasn't yet
  /// resolved. Two things clear it: an `atom-created` event for a
  /// kind=report atom that matches the report's new most-recent
  /// finding (success), or a 30s poll while the detail view is open
  /// that observes `last_error` set with `last_run_at` after dispatch
  /// (failure). A 5-minute stale guard in the detail view clears
  /// optimistic state if neither resolves.
  runningReportIds: Set<string>;

  /// Wall-clock dispatch timestamps per report, epoch ms. Used by the
  /// failure poll to decide whether a stamped `last_run_at` belongs to
  /// the run we just dispatched vs. an earlier one.
  runDispatchedAt: Record<string, number>;

  isLoadingList: boolean;
  loadError: string | null;

  /// Has the atom-created subscription already been set up? Guards
  /// against double-subscription if `fetchAll` is called twice.
  hasSubscription: boolean;

  fetchAll: () => Promise<void>;
  fetchLastFinding: (reportId: string) => Promise<void>;

  /// Fetch a single report by id and merge it into the store. Used by
  /// the detail view on cold-start deep links when the list hasn't
  /// been hydrated yet — calling `fetchAll` would work but is wasteful
  /// for the "open one report directly" path. Returns the report on
  /// success, `null` if the server returned 404 (report deleted).
  fetchOne: (reportId: string) => Promise<Report | null>;

  /// Fetch the full findings history for a report. The detail view
  /// mounts a single load on enter; subsequent live refreshes come
  /// from the AtomCreated event wired in 4c-3.
  fetchFindings: (reportId: string, limit?: number) => Promise<void>;

  /// Fetch the citation count for a finding atom. Idempotent —
  /// returns immediately if already cached. Used by FindingRow's
  /// in-view-effect so we don't paint the whole list up front.
  fetchCitationCount: (atomId: string) => Promise<void>;

  /// Dispatch a manual run. Adds the report to `runningReportIds`,
  /// stamps `runDispatchedAt`, fires `POST /api/reports/:id/run`. The
  /// server responds 202 immediately; we wait for the AtomCreated
  /// event (success) or the failure poll to clear the running state.
  /// Throws on the dispatch itself failing (network / 404); the store
  /// toasts on top.
  runNow: (reportId: string) => Promise<void>;

  /// Clear the running state for one report. Called by the success
  /// detection in the AtomCreated handler, the failure detection in
  /// the detail view's poll, and the 5-minute stale guard.
  clearRunning: (reportId: string) => void;

  /// Create a new report. Returns the created `Report` so the caller
  /// (typically the editor modal) can navigate to it on success.
  /// Throws on failure with a useful message; the store toasts and the
  /// caller can keep its modal open.
  create: (input: CreateReportInput) => Promise<Report>;

  /// Patch an existing report. Returns the merged row from the server.
  /// On failure, throws and leaves the in-memory row untouched.
  update: (id: string, input: UpdateReportInput) => Promise<Report>;

  /// Convenience for the row-level toggle. Optimistic: flips the flag
  /// locally first, reverts on failure. Wired through `update_report`
  /// rather than a dedicated endpoint to keep the transport surface
  /// narrow.
  setEnabled: (id: string, enabled: boolean) => Promise<void>;

  /// Delete a report. Optimistic: removes from the list first, restores
  /// on failure. Findings outlive their producer by design — only the
  /// schedule + definition go away. (The backend already clears the
  /// dashboard's `featured_report_id` if it pointed at this report.)
  delete: (id: string) => Promise<void>;

  reset: () => void;
}

export const useReportsStore = create<ReportsStore>((set, get) => {
  // Module-scope handle for the atom-created unsubscribe so `reset()` can
  // tear it down. Captured in closure rather than store state because it
  // isn't render-relevant.
  let atomCreatedUnsub: (() => void) | null = null;

  return {
    reports: [],
    byId: {},
    lastFindingByReport: {},
    findingsByReport: {},
    citationCountsByAtomId: {},
    runningReportIds: new Set<string>(),
    runDispatchedAt: {},
    isLoadingList: false,
    loadError: null,
    hasSubscription: false,

    fetchAll: async () => {
      set({ isLoadingList: true, loadError: null });
      try {
        const reports = await getTransport().invoke<Report[]>('list_reports');
        const byId: Record<string, Report> = {};
        for (const r of reports) byId[r.id] = r;
        set({ reports, byId, isLoadingList: false });

        // Lazily prime last-finding for every report. Issue requests in
        // parallel; failures degrade to "no excerpt available" without
        // surfacing per-report toasts.
        await Promise.all(reports.map(r => get().fetchLastFinding(r.id)));

        // Wire the live-refresh subscription once. The dashboard
        // BriefingWidget uses the same shape: AtomWithTags flattens, so
        // `kind` lives at the payload top level. When a report finding
        // lands we:
        //   1. Re-prime last-finding for every loaded report (the list
        //      view's italic excerpt + the detail view's findings cache)
        //   2. Refetch each running report's findings, then check which
        //      report's most-recent finding matches the new atom id.
        //      That's the run we completed — clear its running state.
        if (!get().hasSubscription) {
          atomCreatedUnsub = getTransport().subscribe('atom-created', async (payload) => {
            const p = payload as { kind?: string; id?: string } | undefined;
            if (p?.kind !== 'report') return;
            const newAtomId = p.id;

            const state = get();
            const allIds = Object.keys(state.byId);
            const runningIds = Array.from(state.runningReportIds);

            // Re-prime the list view's last-finding cache + every
            // running report's full findings cache. Running reports
            // are a strict subset of allIds, so we issue findings
            // requests for them and last-finding requests for the rest.
            const runningSet = new Set(runningIds);
            const lastFindingPromises = allIds
              .filter(id => !runningSet.has(id))
              .map(id => state.fetchLastFinding(id));
            const findingsPromises = runningIds.map(id => state.fetchFindings(id));
            await Promise.all([...lastFindingPromises, ...findingsPromises]);

            // After cache rehydration, clear any running report whose
            // most-recent finding matches the new atom id.
            if (newAtomId) {
              const after = get();
              for (const id of runningIds) {
                const first = after.findingsByReport[id]?.[0];
                if (first && first.atom.id === newAtomId) {
                  get().clearRunning(id);
                  toast.success('New finding', {
                    description: after.byId[id]?.name,
                  });
                }
              }
            }
          });
          set({ hasSubscription: true });
        }
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        set({ isLoadingList: false, loadError: msg });
        toast.error('Failed to load reports', { description: msg });
      }
    },

    runNow: async (reportId: string) => {
      // Optimistically mark running before the dispatch lands, so the
      // button disables instantly. Even if dispatch fails we revert.
      const dispatchedAt = Date.now();
      set(state => {
        const next = new Set(state.runningReportIds);
        next.add(reportId);
        return {
          runningReportIds: next,
          runDispatchedAt: { ...state.runDispatchedAt, [reportId]: dispatchedAt },
        };
      });
      try {
        await getTransport().invoke('run_report_now', { report_id: reportId });
        // 202 means "accepted, will run". The completion is observed
        // via the AtomCreated subscription (success) or the detail
        // view's failure poll. Nothing else to do here.
      } catch (e) {
        // Dispatch itself failed — revert.
        get().clearRunning(reportId);
        const msg = e instanceof Error ? e.message : String(e);
        toast.error('Failed to dispatch report run', { description: msg });
        throw e;
      }
    },

    clearRunning: (reportId: string) => {
      set(state => {
        if (!state.runningReportIds.has(reportId)) return state;
        const next = new Set(state.runningReportIds);
        next.delete(reportId);
        const { [reportId]: _omit, ...restDispatched } = state.runDispatchedAt;
        return { runningReportIds: next, runDispatchedAt: restDispatched };
      });
    },

    fetchOne: async (reportId: string) => {
      try {
        const report = await getTransport().invoke<Report>('get_report', { report_id: reportId });
        set(state => ({
          // Prepend if not already present, otherwise replace in place.
          // Either way the byId map is the source of truth for the
          // detail view.
          reports: state.byId[reportId]
            ? state.reports.map(r => (r.id === reportId ? report : r))
            : [report, ...state.reports],
          byId: { ...state.byId, [reportId]: report },
        }));
        return report;
      } catch (e) {
        // 404 means the report was deleted; the caller (detail view)
        // surfaces this with a toast and navigates back. Other errors
        // toast directly so they're visible.
        const msg = e instanceof Error ? e.message : String(e);
        if (/not found|404/i.test(msg)) {
          return null;
        }
        toast.error('Failed to load report', { description: msg });
        return null;
      }
    },

    fetchFindings: async (reportId: string, limit = 50) => {
      try {
        const results = await getTransport().invoke<FindingTuple[]>(
          'list_findings_for_report',
          { report_id: reportId, limit }
        );
        const decoded: ReportFindingWithAtom[] = results.map(([finding, atom]) => ({ finding, atom }));
        set(state => ({
          findingsByReport: { ...state.findingsByReport, [reportId]: decoded },
          // Keep the list-view's last-finding cache in sync — it's a
          // strict subset of what we just loaded.
          lastFindingByReport: {
            ...state.lastFindingByReport,
            [reportId]: decoded[0] ?? null,
          },
        }));
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        toast.error('Failed to load findings', { description: msg });
      }
    },

    fetchCitationCount: async (atomId: string) => {
      // Idempotent: skip if already cached.
      if (get().citationCountsByAtomId[atomId] !== undefined) return;
      try {
        const citations = await getTransport().invoke<ReportFindingCitation[]>(
          'list_finding_citations',
          { atom_id: atomId }
        );
        set(state => ({
          citationCountsByAtomId: {
            ...state.citationCountsByAtomId,
            [atomId]: citations.length,
          },
        }));
      } catch (e) {
        // Soft-fail: a missing citation count just hides the badge.
        // Don't toast — could be N concurrent failures during a
        // mass-render and we'd flood the user.
        console.error('[reports] fetchCitationCount failed', atomId, e);
      }
    },

    fetchLastFinding: async (reportId: string) => {
      try {
        const results = await getTransport().invoke<FindingTuple[]>(
          'list_findings_for_report',
          { report_id: reportId, limit: 1 }
        );
        // Server ships JSON 2-tuples (Rust `Vec<(A, B)>` semantics);
        // destructure into the clean shape we cache.
        const first: ReportFindingWithAtom | null = results[0]
          ? { finding: results[0][0], atom: results[0][1] }
          : null;
        set(state => ({
          lastFindingByReport: { ...state.lastFindingByReport, [reportId]: first },
        }));
      } catch (e) {
        // Per-report failure: leave the cache untouched, log, and let the
        // row render without an excerpt. We don't toast — N possible
        // failures would flood the user.
        console.error('[reports] fetchLastFinding failed', reportId, e);
      }
    },

    create: async (input: CreateReportInput) => {
      try {
        const created = await getTransport().invoke<Report>('create_report', input as unknown as Record<string, unknown>);
        // Prepend on success; the list view shows most-recently-created
        // first by default. The next fetchAll will re-canonicalize
        // sort, but in-the-moment ordering should feel snappy.
        set(state => ({
          reports: [created, ...state.reports.filter(r => r.id !== created.id)],
          byId: { ...state.byId, [created.id]: created },
        }));
        toast.success('Report created', { description: created.name });
        return created;
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        toast.error('Failed to create report', { description: msg });
        throw e;
      }
    },

    update: async (id: string, input: UpdateReportInput) => {
      try {
        const merged = await getTransport().invoke<Report>('update_report', {
          report_id: id,
          ...input,
        });
        set(state => ({
          reports: state.reports.map(r => (r.id === id ? merged : r)),
          byId: { ...state.byId, [id]: merged },
        }));
        return merged;
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        toast.error('Failed to update report', { description: msg });
        throw e;
      }
    },

    setEnabled: async (id: string, enabled: boolean) => {
      const prev = get().byId[id];
      if (!prev) return;
      // Optimistic local flip so the row's badge updates instantly.
      const optimistic: Report = { ...prev, enabled };
      set(state => ({
        reports: state.reports.map(r => (r.id === id ? optimistic : r)),
        byId: { ...state.byId, [id]: optimistic },
      }));
      try {
        const merged = await getTransport().invoke<Report>('update_report', {
          report_id: id,
          enabled,
        });
        set(state => ({
          reports: state.reports.map(r => (r.id === id ? merged : r)),
          byId: { ...state.byId, [id]: merged },
        }));
      } catch (e) {
        // Revert. The next fetchAll would heal anyway, but a snappy
        // revert avoids a stuck-toggle perception while the user reads
        // the toast.
        set(state => ({
          reports: state.reports.map(r => (r.id === id ? prev : r)),
          byId: { ...state.byId, [id]: prev },
        }));
        const msg = e instanceof Error ? e.message : String(e);
        toast.error(enabled ? 'Failed to enable report' : 'Failed to pause report', {
          description: msg,
        });
      }
    },

    delete: async (id: string) => {
      const prev = get().reports;
      const prevById = get().byId;
      const target = prevById[id];
      // Optimistic removal.
      set(state => ({
        reports: state.reports.filter(r => r.id !== id),
        byId: Object.fromEntries(Object.entries(state.byId).filter(([k]) => k !== id)),
      }));
      try {
        await getTransport().invoke('delete_report', { report_id: id });
        toast.success('Report deleted', {
          description: target ? `${target.name} — findings remain in your atoms` : undefined,
        });
      } catch (e) {
        // Restore on failure.
        set({ reports: prev, byId: prevById });
        const msg = e instanceof Error ? e.message : String(e);
        toast.error('Failed to delete report', { description: msg });
        throw e;
      }
    },

    reset: () => {
      atomCreatedUnsub?.();
      atomCreatedUnsub = null;
      set({
        reports: [],
        byId: {},
        lastFindingByReport: {},
        findingsByReport: {},
        citationCountsByAtomId: {},
        runningReportIds: new Set<string>(),
        runDispatchedAt: {},
        isLoadingList: false,
        loadError: null,
        hasSubscription: false,
      });
    },
  };
});

File: crates/atomic-server/src/routes/reports.rs
//! Reports REST endpoints. CRUD on report definitions, history view,
//! manual "run now" trigger.
//!
//! Run-now is async: it spawns the runner on the tokio runtime and
//! returns 202 with the report id immediately. The finding atom is
//! observable via `GET /api/reports/:id/findings` (and via the standard
//! atom endpoints once the embedding pipeline finishes).

use crate::db_extractor::Db;
use crate::error::{error_response, ok_or_error, ApiErrorResponse};
use crate::state::{AppState, ServerEvent};
use actix_web::{web, HttpResponse};
use atomic_core::models::{CreateReportRequest, UpdateReportRequest};
use serde::Deserialize;
use utoipa::{IntoParams, ToSchema};

#[utoipa::path(
    get,
    path = "/api/reports",
    responses(
        (status = 200, description = "All reports for this database",
         body = Vec<atomic_core::models::Report>)
    ),
    tag = "reports"
)]
pub async fn list_reports(db: Db) -> HttpResponse {
    ok_or_error(db.0.list_reports().await)
}

#[utoipa::path(
    get,
    path = "/api/reports/{id}",
    responses(
        (status = 200, description = "The requested report",
         body = atomic_core::models::Report),
        (status = 404, description = "No report with this id",
         body = ApiErrorResponse)
    ),
    tag = "reports"
)]
pub async fn get_report(db: Db, path: web::Path<String>) -> HttpResponse {
    let id = path.into_inner();
    match db.0.get_report(&id).await {
        Ok(Some(r)) => HttpResponse::Ok().json(r),
        Ok(None) => HttpResponse::NotFound().json(serde_json::json!({"error": "not found"})),
        Err(e) => error_response(e),
    }
}

#[utoipa::path(
    post,
    path = "/api/reports",
    request_body = CreateReportRequest,
    responses(
        (status = 200, description = "Report created",
         body = atomic_core::models::Report),
        (status = 400, description = "Invalid schedule or timezone",
         body = ApiErrorResponse)
    ),
    tag = "reports"
)]
pub async fn create_report(db: Db, body: web::Json<CreateReportRequest>) -> HttpResponse {
    ok_or_error(db.0.create_report(body.into_inner()).await)
}

#[utoipa::path(
    put,
    path = "/api/reports/{id}",
    request_body = UpdateReportRequest,
    responses(
        (status = 200, description = "Report updated",
         body = atomic_core::models::Report),
        (status = 400, description = "Invalid schedule or timezone",
         body = ApiErrorResponse),
        (status = 404, description = "No report with this id",
         body = ApiErrorResponse)
    ),
    tag = "reports"
)]
pub async fn update_report(
    db: Db,
    path: web::Path<String>,
    body: web::Json<UpdateReportRequest>,
) -> HttpResponse {
    ok_or_error(
        db.0.update_report(&path.into_inner(), body.into_inner())
            .await,
    )
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct SetEnabledBody {
    pub enabled: bool,
}

#[utoipa::path(
    patch,
    path = "/api/reports/{id}/enabled",
    request_body = SetEnabledBody,
    responses(
        (status = 204, description = "Enabled flag updated"),
        (status = 404, description = "No report with this id",
         body = ApiErrorResponse)
    ),
    tag = "reports"
)]
pub async fn set_report_enabled(
    db: Db,
    path: web::Path<String>,
    body: web::Json<SetEnabledBody>,
) -> HttpResponse {
    match db
        .0
        .set_report_enabled(&path.into_inner(), body.enabled)
        .await
    {
        Ok(()) => HttpResponse::NoContent().finish(),
        Err(e) => error_response(e),
    }
}

#[utoipa::path(
    delete,
    path = "/api/reports/{id}",
    responses(
        (status = 204, description = "Report deleted"),
        (status = 404, description = "No report with this id",
         body = ApiErrorResponse)
    ),
    tag = "reports"
)]
pub async fn delete_report(
    db: Db,
    path: web::Path<String>,
    state: web::Data<AppState>,
) -> HttpResponse {
    let id = path.into_inner();
    // Snapshot the featured pointer before deletion so we can detect
    // whether the delete cleared it. `delete_report` clears the pointer
    // internally if it pointed at this report; the broadcast keeps
    // other clients in sync.
    let previously_featured = matches!(
        db.0.get_featured_report_id().await,
        Ok(Some(ref cur)) if cur == &id
    );

    match db.0.delete_report(&id).await {
        Ok(()) => {
            if previously_featured {
                let _ = state
                    .event_tx
                    .send(ServerEvent::DashboardFeaturedChanged { report_id: None });
            }
            HttpResponse::NoContent().finish()
        }
        Err(e) => error_response(e),
    }
}

#[derive(Debug, Deserialize, IntoParams, ToSchema)]
#[into_params(parameter_in = Query)]
pub struct ListFindingsQuery {
    /// Max findings to return (default 50, max 200).
    pub limit: Option<i32>,
}

#[utoipa::path(
    get,
    path = "/api/reports/{id}/findings",
    params(ListFindingsQuery),
    responses(
        (status = 200, description = "Most-recent-first findings joined with atom snippet",
         body = Vec<(atomic_core::models::ReportFinding, atomic_core::models::AtomWithTags)>)
    ),
    tag = "reports"
)]
pub async fn list_findings_for_report(
    db: Db,
    path: web::Path<String>,
    query: web::Query<ListFindingsQuery>,
) -> HttpResponse {
    let limit = query.limit.unwrap_or(50).clamp(1, 200);
    ok_or_error(
        db.0.list_findings_for_report(&path.into_inner(), limit)
            .await,
    )
}

#[utoipa::path(
    get,
    path = "/api/findings/{atom_id}/citations",
    responses(
        (status = 200, description = "Citation rows for a finding atom, ordered by position",
         body = Vec<atomic_core::models::ReportFindingCitation>)
    ),
    tag = "reports"
)]
pub async fn list_finding_citations(db: Db, path: web::Path<String>) -> HttpResponse {
    ok_or_error(db.0.list_citations_for_finding(&path.into_inner()).await)
}

/// Async manual-run response. Includes the report id and a hint at where
/// to poll for results; the actual finding atom shows up via the
/// findings endpoint once the agent loop completes.
#[derive(Debug, serde::Serialize, ToSchema)]
pub struct RunNowResponse {
    pub report_id: String,
    pub status: &'static str,
    pub findings_url: String,
}

#[utoipa::path(
    post,
    path = "/api/reports/{id}/run",
    responses(
        (status = 202, description = "Run dispatched; poll /findings for completion",
         body = RunNowResponse),
        (status = 404, description = "No report with this id",
         body = ApiErrorResponse)
    ),
    tag = "reports"
)]
pub async fn run_report_now(
    db: Db,
    path: web::Path<String>,
    state: web::Data<AppState>,
) -> HttpResponse {
    let id = path.into_inner();
    // Validate the report exists before reporting 202 — otherwise the
    // 404 would be deferred to the background task and the caller would
    // see a successful dispatch for a nonexistent id.
    match db.0.get_report(&id).await {
        Ok(Some(_)) => {}
        Ok(None) => {
            return HttpResponse::NotFound().json(serde_json::json!({"error": "report not found"}))
        }
        Err(e) => return error_response(e),
    }

    let core = db.0.clone();
    let id_owned = id.clone();
    let event_tx = state.event_tx.clone();
    tokio::spawn(async move {
        match core.run_report_now(&id_owned).await {
            Ok(outcome) => {
                tracing::info!(
                    report_id = %id_owned,
                    outcome = ?outcome,
                    "[reports/run-now] complete"
                );
                // Broadcast `atom-created` on success so the dashboard
                // widget refreshes live. Matches what the scheduled-run
                // loop emits in `atomic-server::main`.
                if let atomic_core::reports::RunOutcome::Succeeded { finding_atom_id } = outcome {
                    if let Ok(Some(atom)) = core.get_atom(&finding_atom_id).await {
                        let _ = event_tx.send(ServerEvent::AtomCreated { atom });
                    }
                }
            }
            Err(e) => tracing::error!(
                report_id = %id_owned,
                error = %e,
                "[reports/run-now] failed"
            ),
        }
    });

    HttpResponse::Accepted().json(RunNowResponse {
        report_id: id.clone(),
        status: "dispatched",
        findings_url: format!("/api/reports/{id}/findings"),
    })
}
